File: tests/sdk/nodejs/test/observe-govern-labs.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateInstancesCommand,
  DeleteInstanceCommand,
  GetInstanceCommand,
  LightsailClient,
} from "@aws-sdk/client-lightsail";
import {
  AutoScalingClient,
  CreateAutoScalingGroupCommand,
  CreateLaunchConfigurationCommand,
  DeleteAutoScalingGroupCommand,
  DeleteLaunchConfigurationCommand,
  DescribeAutoScalingGroupsCommand,
  SetDesiredCapacityCommand,
} from "@aws-sdk/client-auto-scaling";
import {
  CreateApplicationCommand,
  CreateEnvironmentCommand,
  DeleteApplicationCommand,
  ElasticBeanstalkClient,
  TerminateEnvironmentCommand,
} from "@aws-sdk/client-elastic-beanstalk";
import {
  endpoint,
  region,
  requireReady,
  signedFetch,
  uniquePrefix,
} from "../lib/helpers.mjs";

function baseClientConfig() {
  return {
    region: region(),
    endpoint: endpoint(),
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID || "AKIAROOTEXAMPLE01",
      secretAccessKey:
        process.env.AWS_SECRET_ACCESS_KEY ||
        "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
    },
  };
}

test("Lightsail create get delete", async (t) => {
  if (!(await requireReady(t))) return;
  const c = new LightsailClient(baseClientConfig());
  const name = `${uniquePrefix()}-ls`;
  await c.send(
    new CreateInstancesCommand({
      instanceNames: [name],
      availabilityZone: "us-east-1a",
      blueprintId: "ubuntu_22_04",
      bundleId: "nano_3_0",
    }),
  );
  t.after(async () => {
    try {
      await c.send(new DeleteInstanceCommand({ instanceName: name }));
    } catch {
      /* ignore */
    }
  });
  const got = await c.send(new GetInstanceCommand({ instanceName: name }));
  assert.equal(got.instance?.name, name);
});

test("Auto Scaling SetDesiredCapacity yields InstanceIds", async (t) => {
  if (!(await requireReady(t))) return;
  const c = new AutoScalingClient(baseClientConfig());
  const prefix = uniquePrefix();
  const lc = `${prefix}-lc`;
  const asg = `${prefix}-asg`;
  await c.send(
    new CreateLaunchConfigurationCommand({
      LaunchConfigurationName: lc,
      ImageId: "ami-12345678",
      InstanceType: "t3.micro",
    }),
  );
  t.after(async () => {
    try {
      await c.send(
        new DeleteAutoScalingGroupCommand({
          AutoScalingGroupName: asg,
          ForceDelete: true,
        }),
      );
    } catch {
      /* ignore */
    }
    try {
      await c.send(
        new DeleteLaunchConfigurationCommand({ LaunchConfigurationName: lc }),
      );
    } catch {
      /* ignore */
    }
  });
  await c.send(
    new CreateAutoScalingGroupCommand({
      AutoScalingGroupName: asg,
      LaunchConfigurationName: lc,
      MinSize: 1,
      MaxSize: 3,
      DesiredCapacity: 1,
      AvailabilityZones: ["us-east-1a"],
    }),
  );
  await c.send(
    new SetDesiredCapacityCommand({
      AutoScalingGroupName: asg,
      DesiredCapacity: 2,
    }),
  );
  const desc = await c.send(
    new DescribeAutoScalingGroupsCommand({ AutoScalingGroupNames: [asg] }),
  );
  assert.equal(desc.AutoScalingGroups?.[0]?.DesiredCapacity, 2);
  const members = desc.AutoScalingGroups?.[0]?.Instances || [];
  assert.equal(members.length, 2);
  for (const m of members) {
    assert.match(String(m.InstanceId || ""), /^i-/);
    const ls = String(m.LifecycleState || "");
    assert.ok(
      ls === "Pending" || ls === "InService",
      `LifecycleState=${ls} (Pending OK without engine)`,
    );
  }
});

test("Elastic Beanstalk Ready environment", async (t) => {
  if (!(await requireReady(t))) return;
  const c = new ElasticBeanstalkClient(baseClientConfig());
  const prefix = uniquePrefix();
  const app = `${prefix}-app`;
  const env = `${prefix}-env`;
  await c.send(new CreateApplicationCommand({ ApplicationName: app }));
  t.after(async () => {
    try {
      await c.send(new TerminateEnvironmentCommand({ EnvironmentName: env }));
    } catch {
      /* ignore */
    }
    try {
      await c.send(
        new DeleteApplicationCommand({
          ApplicationName: app,
          TerminateEnvByForce: true,
        }),
      );
    } catch {
      /* ignore */
    }
  });
  const created = await c.send(
    new CreateEnvironmentCommand({
      ApplicationName: app,
      EnvironmentName: env,
    }),
  );
  assert.equal(created.Status, "Ready");
});

test("Backup vault plan job recovery point", async (t) => {
  if (!(await requireReady(t))) return;
  const prefix = uniquePrefix();
  const vault = `${prefix}-vault`.toLowerCase();

  const createVault = await signedFetch(
    "backup",
    "PUT",
    `/backup-vaults/${encodeURIComponent(vault)}`,
    JSON.stringify({}),
    "application/json",
  );
  assert.equal(createVault.status, 200, await createVault.text());
  t.after(async () => {
    await signedFetch(
      "backup",
      "DELETE",
      `/backup-vaults/${encodeURIComponent(vault)}`,
      null,
      "",
    );
  });

  const createPlan = await signedFetch(
    "backup",
    "PUT",
    "/backup/plans/",
    JSON.stringify({
      BackupPlan: {
        BackupPlanName: `${prefix}-plan`,
        Rules: [
          {
            RuleName: "daily",
            TargetBackupVaultName: vault,
            ScheduleExpression: "cron(0 5 ? * * *)",
          },
        ],
      },
    }),
    "application/json",
  );
  const planBody = await createPlan.text();
  assert.equal(createPlan.status, 200, planBody);
  assert.match(planBody, /BackupPlanId/);

  const start = await signedFetch(
    "backup",
    "PUT",
    "/backup-jobs",
    JSON.stringify({
      BackupVaultName: vault,
      ResourceArn: "arn:aws:s3:::lab-bucket",
      IamRoleArn: "arn:aws:iam::000000000001:role/Backup",
    }),
    "application/json",
  );
  const startBody = await start.text();
  assert.equal(start.status, 200, startBody);
  assert.match(startBody, /RecoveryPointArn/);

  const list = await signedFetch(
    "backup",
    "GET",
    `/backup-vaults/${encodeURIComponent(vault)}/recovery-points/`,
    null,
    "",
  );
  const listBody = await list.text();
  assert.equal(list.status, 200, listBody);
  assert.match(listBody, /S3/);
});

File: tests/sdk/go/observe_govern_labs_test.go
package sdk_test

import (
	"context"
	"encoding/json"
	"net/url"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
	"github.com/aws/aws-sdk-go-v2/service/elasticbeanstalk"
	"github.com/aws/aws-sdk-go-v2/service/lightsail"
)

func TestLightsailCreateGetDelete(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	c := lightsail.NewFromConfig(cfg, func(o *lightsail.Options) {
		o.BaseEndpoint = aws.String(endpoint())
	})
	ctx := context.Background()
	name := uniquePrefix(t) + "-ls"

	if _, err := c.CreateInstances(ctx, &lightsail.CreateInstancesInput{
		InstanceNames:    []string{name},
		AvailabilityZone: aws.String("us-east-1a"),
		BlueprintId:      aws.String("ubuntu_22_04"),
		BundleId:         aws.String("nano_3_0"),
	}); err != nil {
		t.Fatalf("CreateInstances: %v", err)
	}
	t.Cleanup(func() {
		_, _ = c.DeleteInstance(ctx, &lightsail.DeleteInstanceInput{InstanceName: aws.String(name)})
	})

	got, err := c.GetInstance(ctx, &lightsail.GetInstanceInput{InstanceName: aws.String(name)})
	if err != nil {
		t.Fatalf("GetInstance: %v", err)
	}
	if got.Instance == nil || aws.ToString(got.Instance.Name) != name {
		t.Fatalf("instance=%+v", got.Instance)
	}
}

func TestAutoScalingGroupDesiredCapacity(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	c := autoscaling.NewFromConfig(cfg, func(o *autoscaling.Options) {
		o.BaseEndpoint = aws.String(endpoint())
	})
	ctx := context.Background()
	prefix := uniquePrefix(t)
	lc := prefix + "-lc"
	asg := prefix + "-asg"

	if _, err := c.CreateLaunchConfiguration(ctx, &autoscaling.CreateLaunchConfigurationInput{
		LaunchConfigurationName: aws.String(lc),
		ImageId:                 aws.String("ami-12345678"),
		InstanceType:            aws.String("t3.micro"),
	}); err != nil {
		t.Fatalf("CreateLaunchConfiguration: %v", err)
	}
	t.Cleanup(func() {
		_, _ = c.DeleteAutoScalingGroup(ctx, &autoscaling.DeleteAutoScalingGroupInput{
			AutoScalingGroupName: aws.String(asg), ForceDelete: aws.Bool(true),
		})
		_, _ = c.DeleteLaunchConfiguration(ctx, &autoscaling.DeleteLaunchConfigurationInput{
			LaunchConfigurationName: aws.String(lc),
		})
	})

	if _, err := c.CreateAutoScalingGroup(ctx, &autoscaling.CreateAutoScalingGroupInput{
		AutoScalingGroupName:    aws.String(asg),
		LaunchConfigurationName: aws.String(lc),
		MinSize:                 aws.Int32(1),
		MaxSize:                 aws.Int32(3),
		DesiredCapacity:         aws.Int32(1),
		AvailabilityZones:       []string{"us-east-1a"},
	}); err != nil {
		t.Fatalf("CreateAutoScalingGroup: %v", err)
	}

	if _, err := c.SetDesiredCapacity(ctx, &autoscaling.SetDesiredCapacityInput{
		AutoScalingGroupName: aws.String(asg),
		DesiredCapacity:      aws.Int32(2),
	}); err != nil {
		t.Fatalf("SetDesiredCapacity: %v", err)
	}

	desc, err := c.DescribeAutoScalingGroups(ctx, &autoscaling.DescribeAutoScalingGroupsInput{
		AutoScalingGroupNames: []string{asg},
	})
	if err != nil || len(desc.AutoScalingGroups) != 1 {
		t.Fatalf("DescribeAutoScalingGroups: %+v err=%v", desc, err)
	}
	if aws.ToInt32(desc.AutoScalingGroups[0].DesiredCapacity) != 2 {
		t.Fatalf("desired=%v", desc.AutoScalingGroups[0].DesiredCapacity)
	}
	members := desc.AutoScalingGroups[0].Instances
	if len(members) != 2 {
		t.Fatalf("expected 2 Instances after SetDesiredCapacity, got %+v", members)
	}
	for i, m := range members {
		id := aws.ToString(m.InstanceId)
		if !strings.HasPrefix(id, "i-") {
			t.Fatalf("member[%d] InstanceId=%q", i, id)
		}
		ls := string(m.LifecycleState)
		if ls != "Pending" && ls != "InService" {
			t.Fatalf("member[%d] LifecycleState=%q (Pending OK without engine)", i, ls)
		}
	}
}

func TestBeanstalkApplicationEnvironment(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	c := elasticbeanstalk.NewFromConfig(cfg, func(o *elasticbeanstalk.Options) {
		o.BaseEndpoint = aws.String(endpoint())
	})
	ctx := context.Background()
	prefix := uniquePrefix(t)
	app := prefix + "-app"
	env := prefix + "-env"

	if _, err := c.CreateApplication(ctx, &elasticbeanstalk.CreateApplicationInput{
		ApplicationName: aws.String(app),
	}); err != nil {
		t.Fatalf("CreateApplication: %v", err)
	}
	t.Cleanup(func() {
		_, _ = c.TerminateEnvironment(ctx, &elasticbeanstalk.TerminateEnvironmentInput{
			EnvironmentName: aws.String(env),
		})
		_, _ = c.DeleteApplication(ctx, &elasticbeanstalk.DeleteApplicationInput{
			ApplicationName:     aws.String(app),
			TerminateEnvByForce: aws.Bool(true),
		})
	})

	created, err := c.CreateEnvironment(ctx, &elasticbeanstalk.CreateEnvironmentInput{
		ApplicationName: aws.String(app),
		EnvironmentName: aws.String(env),
	})
	if err != nil {
		t.Fatalf("CreateEnvironment: %v", err)
	}
	if string(created.Status) != "Ready" {
		t.Fatalf("status=%q", string(created.Status))
	}
}

func TestBackupVaultPlanJob(t *testing.T) {
	requireReady(t)
	prefix := uniquePrefix(t)
	vault := strings.ToLower(prefix + "-vault")

	status, body := signedHTTP(t, "backup", "PUT", "/backup-vaults/"+url.PathEscape(vault), mustJSONBytes(t, map[string]any{}), "application/json")
	if status != 200 {
		t.Fatalf("CreateBackupVault status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _ = signedHTTP(t, "backup", "DELETE", "/backup-vaults/"+url.PathEscape(vault), nil, "")
	})

	status, body = signedHTTP(t, "backup", "PUT", "/backup/plans/", mustJSONBytes(t, map[string]any{
		"BackupPlan": map[string]any{
			"BackupPlanName": prefix + "-plan",
			"Rules": []map[string]any{
				{"RuleName": "daily", "TargetBackupVaultName": vault, "ScheduleExpression": "cron(0 5 ? * * *)"},
			},
		},
	}), "application/json")
	if status != 200 || !strings.Contains(string(body), "BackupPlanId") {
		t.Fatalf("CreateBackupPlan status=%d body=%s", status, body)
	}
	var plan map[string]any
	_ = json.Unmarshal(body, &plan)
	if id, _ := plan["BackupPlanId"].(string); id != "" {
		t.Cleanup(func() {
			_, _ = signedHTTP(t, "backup", "DELETE", "/backup/plans/"+url.PathEscape(id), nil, "")
		})
	}

	status, body = signedHTTP(t, "backup", "PUT", "/backup-jobs", mustJSONBytes(t, map[string]any{
		"BackupVaultName": vault,
		"ResourceArn":     "arn:aws:s3:::lab-bucket",
		"IamRoleArn":      "arn:aws:iam::000000000001:role/Backup",
	}), "application/json")
	if status != 200 || !strings.Contains(string(body), "RecoveryPointArn") {
		t.Fatalf("StartBackupJob status=%d body=%s", status, body)
	}

	status, body = signedHTTP(t, "backup", "GET", "/backup-vaults/"+url.PathEscape(vault)+"/recovery-points/", nil, "")
	if status != 200 || !strings.Contains(string(body), "S3") {
		t.Fatalf("ListRecoveryPoints status=%d body=%s", status, body)
	}
}

func mustJSONBytes(t *testing.T, v any) []byte {
	t.Helper()
	b, err := json.Marshal(v)
	if err != nil {
		t.Fatal(err)
	}
	return b
}
